crates/ruff_linter/src/rules/flake8_pyi/rules/bad_generator_return_type.rs

```diff
@@ -0,0 +1,191 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_python_ast::helpers::map_subscript;
+use ruff_python_ast::identifier::Identifier;
+use ruff_python_semantic::{ScopeKind, SemanticModel};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for simple `__iter__` methods that return `Generator`, and for
+/// simple `__aiter__` methods that return `AsyncGenerator`.
+///
+/// ## Why is this bad?
+/// Using `(Async)Iterator` for these methods is simpler and more elegant. More
+/// importantly, it also reflects the fact that the precise kind of iterator
+/// returned from an `__iter__` method is usually an implementation detail that
+/// could change at any time. Type annotations help define a contract for a
+/// function; implementation details should not leak into that contract.
+///
+/// For example:
+/// ```python
+/// from collections.abc import AsyncGenerator, Generator
+/// from typing import Any
+///
+///
+/// class CustomIterator:
+///     def __iter__(self) -> Generator:
+///         yield from range(42)
+///
+///
+/// class CustomIterator2:
+///     def __iter__(self) -> Generator[str, Any, None]:
+///         yield from "abcdefg"
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from collections.abc import Iterator
+///
+///
+/// class CustomIterator:
+///     def __iter__(self) -> Iterator:
+///         yield from range(42)
+///
+///
+/// class CustomIterator2:
+///     def __iter__(self) -> Iterator[str]:
+///         yield from "abdefg"
+/// ```
+#[violation]
+pub struct GeneratorReturnFromIterMethod {
+    better_return_type: String,
+    method_name: String,
+}
+
+impl Violation for GeneratorReturnFromIterMethod {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let GeneratorReturnFromIterMethod {
+            better_return_type,
+            method_name,
+        } = self;
+        format!("Use `{better_return_type}` as the return value for simple `{method_name}` methods")
+    }
+}
+
+/// PYI058
+pub(crate) fn bad_generator_return_type(
+    checker: &mut Checker,
+    stmt: &ast::Stmt,
+    is_async: bool,
+    name: &str,
+    returns: Option<&ast::Expr>,
+    parameters: &ast::Parameters,
+    body: &Vec<ast::Stmt>,
+) {
+    if is_async {
+        return;
+    }
+
+    let better_return_type = match name {
+        "__iter__" => "Iterator",
+        "__aiter__" => "AsyncIterator",
+        _ => return,
+    };
+
+    let semantic = checker.semantic();
+
+    if !matches!(checker.semantic().current_scope().kind, ScopeKind::Class(_)) {
```

Comment:
Nit: can use `semantic` here since you already assigned it as `let semantic = checker.semantic();` above.

Reply:
As a handy tricky, you can also do `semantic.current_scope().kind.is_class()`. (This isn't a built-in Rust thing, we have a macro that adds these `is_*` methods to the enum.)